File: 94.二叉树的中序遍历.java
/*
 * @lc app=leetcode.cn id=94 lang=java
 *
 * [94] 二叉树的中序遍历
 */
import java.util.*;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root==null)  return res;
        Stack<Object> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            Object o = stack.pop();
            if (o instanceof TreeNode) {
                TreeNode node = (TreeNode) o;
                // 因为中序遍历是左节点--根节点--右节点
                // 即出栈顺序为左节点--根节点--右节点，入栈顺序相反
                if (node.right != null) stack.push(node.right);
                stack.push(node.val); 
                if (node.left != null) stack.push(node.left); 
            } else {
                res.add((int) o);
            }
       }
       return res;  
    }
}
// @lc code=end

File: 94.二叉树的中序遍历.py
#
# @lc app=leetcode.cn id=94 lang=python3
#
# [94] 二叉树的中序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        """Recursive approach"""
        if not root:
            return []
        res = []
        def inorder(node):
            if node.left is not None:
                inorder(node.left)
            res.append(node.val)
            if node.right is not None:
                inorder(node.right)
        inorder(root)
        return res

    def inorderTraversal(self, root: TreeNode) -> List[int]:
        """Iterative approach with a stack (FILO or LIFO)"""
        if not root:
            return []
        stack, res = [root], []
        while stack:
            i = stack.pop()
            if isinstance(i, TreeNode):
                stack.extend([i.right, i.val, i.left])  # note the order of nodes in extension list
            else:   # elif isinstance(i, int):
                res.append(i)
        return res

# @lc code=end

